From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 12 Feb 2024 20:10:08 +0100
Subject: [PATCH] Ignore system files and directories

Ignore technical files and directories, such as `.git` for Git repositories or `.DS_Store` on macOS file systems,
when scanning for files and directories, such as when reading a data pack from a plain directory.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/server/packs/PathPackResources.java b/src/main/java/net/minecraft/server/packs/PathPackResources.java
index 3edd14ce90edf98798b89921ad18547809de5d2c..e7b16305b1fdcbd5bc866ab87ab77c2e090c3353 100644
--- a/src/main/java/net/minecraft/server/packs/PathPackResources.java
+++ b/src/main/java/net/minecraft/server/packs/PathPackResources.java
@@ -108,6 +108,11 @@ public class PathPackResources extends AbstractPackResources {
         try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
             for(Path path2 : directoryStream) {
                 String string = path2.getFileName().toString();
+                // Fiddle start - ignore system files and directories
+                if (org.fiddlemc.fiddle.pack.read.SystemFilesAndDirectories.isSystemFileOrDirectory(string)) {
+                    continue;
+                }
+                // Fiddle end - ignore system files and directories
                 // Paper start - Improve logging and errors
                 if (!Files.isDirectory(path2)) {
                     LOGGER.error("Invalid directory entry: {} in {}.", string, this.root, new java.nio.file.NotDirectoryException(string));
diff --git a/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java b/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java
index 36bc04fdb3da81283a1a85f7c649b0ceddd163ab..360aaac678facbce7d8d44ca48c168301e1617c4 100644
--- a/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java
+++ b/src/main/java/net/minecraft/server/packs/repository/FolderRepositorySource.java
@@ -65,6 +65,12 @@ public class FolderRepositorySource implements RepositorySource {
         try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
             for(Path path2 : directoryStream) {
                 try {
+                    // Fiddle start - ignore system files and directories
+                    var path2Filename = path2.getFileName();
+                    if (path2Filename != null && org.fiddlemc.fiddle.pack.read.SystemFilesAndDirectories.isSystemFileOrDirectory(path2Filename.toString())) {
+                        continue;
+                    }
+                    // Fiddle end - ignore system files and directories
                     List<ForbiddenSymlinkInfo> list = new ArrayList<>();
                     Pack.ResourcesSupplier resourcesSupplier = folderPackDetector.detectPackResources(path2, list);
                     if (!list.isEmpty()) {
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/read/SystemFilesAndDirectories.java b/src/main/java/org/fiddlemc/fiddle/pack/read/SystemFilesAndDirectories.java
new file mode 100644
index 0000000000000000000000000000000000000000..18bdb753c86dc9c1b044239a5e1b8244c7d5ac8f
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/read/SystemFilesAndDirectories.java
@@ -0,0 +1,38 @@
+// Fiddle - ignore system files and directories
+
+package org.fiddlemc.fiddle.pack.read;
+
+public class SystemFilesAndDirectories {
+
+    private final static String[] FILES = {
+        ".DS_Store",
+        "Thumbs.db",
+    };
+
+    private final static String[] DIRECTORIES = {
+        ".git",
+    };
+
+    public static boolean isSystemFile(String filename) {
+        for (var systemDirectory : FILES) {
+            if (systemDirectory.equals(filename)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static boolean isSystemDirectory(String filename) {
+        for (var systemDirectory : DIRECTORIES) {
+            if (systemDirectory.equals(filename)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static boolean isSystemFileOrDirectory(String filename) {
+        return isSystemFile(filename) || isSystemDirectory(filename);
+    }
+
+}
